File: backend/Enrich.BLL/Services/QuizService.cs
using Enrich.BLL.DTOs;
using Enrich.BLL.Interfaces;
using Enrich.DAL.Entities;
using Enrich.DAL.Interfaces;
using Microsoft.Extensions.Logging;

namespace Enrich.BLL.Services
{
    public class QuizService(
        IQuizRepository quizRepository,
        IFlashcardRepository flashcardRepository,
        ILogger<QuizService> logger) : IQuizService
    {
        public async Task<IEnumerable<FlashcardDTO>> GetRandomSavedFlashcardsAsync(int userId, int count, CancellationToken cancellationToken = default)
        {
            var personalCollection = await flashcardRepository.GetPersonalCollectionAsync(userId, cancellationToken);

            // Randomize and take 'count'
            var randomSet = personalCollection
                .OrderBy(_ => Guid.NewGuid())
                .Take(count)
                .Select(f => new FlashcardDTO
                {
                    Id = f.Id,
                    Word = f.Word,
                    Translation = f.Translation,
                    Transcription = f.Transcription,
                    Meaning = f.Meaning,
                    PartOfSpeech = f.PartOfSpeech,
                    Example = f.Example,
                    DifficultyLevel = f.DifficultyLevel,
                    CreatedAt = f.CreatedAt,
                    IsSaved = true,
                    IsPersonal = f.CreatedById == userId
                });

            return randomSet;
        }

        public async Task<QuizAttemptDTO> SaveQuizResultAsync(int userId, QuizResultDTO result, CancellationToken cancellationToken = default)
        {
            var correctCount = result.Answers.Count(a => a.IsKnown);
            var totalCount = result.Answers.Count;
            var scorePercentage = totalCount > 0 ? (int)((double)correctCount / totalCount * 100) : 0;

            var attempt = new QuizAttempt
            {
                UserId = userId,
                StartedAt = result.StartedAt,
                FinishedAt = result.FinishedAt,
                ScorePercentage = scorePercentage,
                Answers = result.Answers.Select(a => new QuizAnswer
                {
                    FlashcardId = a.FlashcardId,
                    IsKnown = a.IsKnown
                }).ToList()
            };

            var savedAttempt = await quizRepository.AddAttemptAsync(attempt, cancellationToken);

            logger.LogInformation("User {UserId} completed quiz with score {Score}%", userId, scorePercentage);

            return new QuizAttemptDTO
            {
                Id = savedAttempt.Id,
                StartedAt = savedAttempt.StartedAt,
                FinishedAt = savedAttempt.FinishedAt,
                ScorePercentage = savedAttempt.ScorePercentage
            };
        }

        public async Task<IEnumerable<QuizAttemptDTO>> GetUserQuizHistoryAsync(int userId, CancellationToken cancellationToken = default)
        {
            var history = await quizRepository.GetUserHistoryAsync(userId, cancellationToken);
            return history.Select(h => new QuizAttemptDTO
            {
                Id = h.Id,
                StartedAt = h.StartedAt,
                FinishedAt = h.FinishedAt,
                ScorePercentage = h.ScorePercentage
            });
        }
    }
}

File: mobile/src/services/quizService.ts
import { api } from './api';

export interface Flashcard {
  id: number;
  word: string;
  translation?: string;
  transcription?: string;
  meaning?: string;
  partOfSpeech?: string;
  example?: string;
  difficultyLevel?: string;
  isSaved: boolean;
  isPersonal: boolean;
}

export interface QuizAnswer {
  flashcardId: number;
  isKnown: boolean;
}

export interface QuizResult {
  startedAt: string;
  finishedAt: string;
  answers: QuizAnswer[];
}

export interface QuizAttempt {
  id: number;
  startedAt: string;
  finishedAt: string;
  scorePercentage: number;
}

export const quizService = {
  generateQuiz: async (count: number = 10): Promise<Flashcard[]> => {
    const response = await api.get<Flashcard[]>(`Quiz/generate?count=${count}`);
    return response.data;
  },

  submitResult: async (result: QuizResult): Promise<QuizAttempt> => {
    const response = await api.post<QuizAttempt>('Quiz/submit', result);
    return response.data;
  },

  getHistory: async (): Promise<QuizAttempt[]> => {
    const response = await api.get<QuizAttempt[]>('Quiz/history');
    return response.data;
  },
};
